ls1intum/Artemis
# [Bugfix] Show proper error messages in exam management

### Checklist

### Motivation and Context


When actions in the student exam management (like generating student exams, missing student exams, assessing unsubmitted student exams, ...) fail the client does not show a proper error message, but instead only shows "Bad Request" or something similar. Especially in the case of student exam generation the user should get detailed information about the failure.

### Description

The previous error handling only defaulted to generic error handling, which simply shows a translated error message depending on the response status of the query. I added a method, which is always called when one of related queries fails. This new method shows a translated string specific to the failed action and gives some reasoning (if provided in the request response). Otherwise it falls back to the original generic error handling.

### Steps for Testing


1. Create a new exam and give it a wrong configuration (e.g. set number of exercise groups to 2, but only add 1 exercise group)
2. Add some students to the exam (Exam Management -> Students -> Import students or Register course students)
2. Navigate to Exam Management -> Student Exams
3. Click on generate individual exams
4. Assert that the alert shows the specific reasoning why the action failed
5. Also try to cause errors in the other actions on this site (generate missing individual exams, prepare exercise start, unlock all repositories, lock all repositories, evaluate quizzes, assess unsubmitted student exams) and verify the alert messages

Note:
Other actions (except for generating individual exams or generating missing individual exams) do not provide that specific reasoning, so in this case only a translated message specific to that action is shown with the error.message appended if available. Otherwise the old error handling logic applies (this happens for example if the user has no internet connection and the request fails).



### Test Coverage



student-exams.component.ts 96%

### Screenshots



**Before**
![image](https://user-images.githubusercontent.com/12861668/125599096-bf37a370-baa2-4295-a9de-4b68b9aaa437.png)

**After**
![image](https://user-images.githubusercontent.com/12861668/125592519-70c09475-414b-4f19-b8dd-1392ff54c55a.png)
